pkg/spoke/addon/configuration.go

```diff
@@ -0,0 +1,159 @@
+package addon
+
+import (
+	"crypto/sha256"
+	"crypto/x509/pkix"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	addonv1alpha1 "github.com/open-cluster-management/api/addon/v1alpha1"
+	certificatesv1beta1 "k8s.io/api/certificates/v1beta1"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const (
+	installationNamespaceAnnotation = "installNamespace"
+	registrationAnnotation          = "registrations"
+)
+
+// addOnConfig contains meta information of an addon
+type addOnConfig struct {
+	AddOnName             string
+	InstallationNamespace string
+}
+
+// registrationConfig contains necessary information for addon registration
+type registrationConfig struct {
+	AddOnName             string      `json:"addOnName,omitempty"`
+	InstallationNamespace string      `json:"installNamespace,omitempty"`
```

Comment:
why are there json tags here?  In kube we try to separate serialized types from unserialized types.  We do this because serialized types represent a guarantee to a user. Unserialized types have no guarantees.

Reply:
Json tags are added on fields because,
- I need them to unserialized registrationConfs from annotation of `ManagedClusterAddOn`.
- I serialize `registrationConf` into json string and then calculate the hash of `registrationConf`. `registrationConfig` is not comparable because `registrationConfig.Subject.Organization` is an array. Hash is used to compare registrationConfigs and also can be used as key of map.